### PR TITLE
Fixes #9857 - facter extensions now compatible with 1.8+

### PR DESCRIPTION
**Pull request by lzap**
Because of this patch that is in core develop and 1.8:

https://github.com/theforeman/foreman/commit/43c4bd72e4646cc9b2b4cb0533e84e08825eadda

Our facter extensions were no longer called. Since we had poor test coverage,
we found out too late. Puppet extensions were no longer active essentially.

This patch corrects this and fails early when extensions are not properly
included. Also test coverage is now better with more edge cases. We raise more
failures and also if `discovery_bootif` fact is missing/incorrect, we
explicitly error out instead of trying to fallback. This fact is now burnt into
the image and required to successfully discover a host.

Tests will fail because of:
https://github.com/theforeman/foreman_discovery/pull/191

### CODE REVIEW

**Comment by stbenjam** [comment]
[test]

**Comment by lzap** [comment]
Pushed updates to unit tests. So this patch basically fix two problems:
- Our facter extensions were not effective anymore and we did not realize that. Luckily, in most cases facter reports the primary interface via `macaddress` fact correctly and the fallback mechanism did work properly. But this fixes that and determines the primary interface now only via `discovery_bootif` fact and if that's missing we error out.
- The `ignore_puppet_facts_for_provisioning` setting was giving errors because the new 1.8 networking code skipped interfaces parsing, therefore `host.primary_interface` was nil, therefore `Subnet.for(host.primary_interface.ip)` was failing. Discovery now ignores this setting so it works on all deployments.

Also what's in the box:
- Moved org/location determination code to better place and now we default to Subnet's first org/loc when `discovery_organization/location` is not set and only if that's not found we default to the first Org/Loc of the whole deployment. This allows user at least to set Org/Loc via Subnets (if they keep one subnet = one org/loc rule). Our plans are to improve this area in the next release(s).
- Improved unit tests to prevent from above errors.

Ready for testing! Thanks.

**Comment by bilco105** [comment]
This works perfectly for us now @lzap - thanks!

**Comment by lzap** [comment]
Thanks, @stbenjam for some reason all tests fail in jenkins, but are green on my system. Would you mind to run them manually and check? Could be some Setting of mine or something:

```
foreman# rake test:discovery
```

Thanks!

**Comment by lzap** [comment]
Pushed fix, foreman core tests were failing, not discovery ones! Ready for review.

File: app/lib/puppet_fact_parser_extensions.rb [hunk 1]
```diff
@@ -2,42 +2,28 @@ module PuppetFactParserExtensions
...
-    # In Foreman 1.8 these two methods have been removed, we could reuse discovery_mac_fact_name in suggested_primary_interface(host) if needed
-    if instance_methods.include?(:primary_interface)
-      alias_method_chain :primary_interface, :discovery_fact
-    end
-
-    if instance_methods.include?(:ip)
-      alias_method_chain :ip, :discovery_fact
+    begin
+      raise "method: suggested_primary_interface" unless instance_methods.include?(:suggested_primary_interface)
+      raise "method: parse_interfaces?" unless instance_methods.include?(:parse_interfaces?)
```
Comment (stbenjam) [inline app/lib/puppet_fact_parser_extensions.rb:7]: Is this required? Presumably, we're enforcing a version of Foreman in the engine.rb that has these methods.

File: test/unit/host_discovered_test.rb [hunk 2]
```diff
@@ -24,12 +23,43 @@ class HostDiscoveredTest < ActiveSupport::TestCase
     assert Host::Discovered.find_by_name('mace41f13cc3658')
   end
 
+  test "should setup subnet" do
+    raw = parse_json_fixture('/facts.json')
+    subnet = FactoryGirl.create(:subnet)
+    Subnet.expects(:subnet_for).with('10.35.27.3').returns(subnet)
+    subnet.stubs(:organizations).returns([Organization.first])
+    subnet.stubs(:locations).returns([Location.first])
```
Comment (stbenjam) [inline test/unit/host_discovered_test.rb]: Why stub the org/loc? This could do, I think: 

``` ruby
subnet = FactoryGirl.create(:subnet, :organizations => [Organization.first], :locations => [Location.first])
```

File: test/unit/host_discovered_test.rb [hunk 3]
```diff
@@ -24,12 +23,43 @@ class HostDiscoveredTest < ActiveSupport::TestCase
...
+                       :category => 'Setting::Discovered')
+    FactoryGirl.create(:setting,
+                       :name => 'discovery_location',
+                       :value => loc.name,
+                       :category => 'Setting::Discovered')
+    raw = parse_json_fixture('/facts.json')
+    subnet = FactoryGirl.create(:subnet)
+    Subnet.expects(:subnet_for).with('10.35.27.3').returns(subnet)
+    subnet.stubs(:organizations).returns([org])
+    subnet.stubs(:locations).returns([loc])
```
Comment (stbenjam) [inline test/unit/host_discovered_test.rb]: Ditto

File: app/lib/puppet_fact_parser_extensions.rb (hunk 1 above)
Comment (lzap) [inline app/lib/puppet_fact_parser_extensions.rb:7]: Yes, we enforce the version, but when those methods were refactored and disappeared, we haven't realized this at all. I'd rather be explicit. Or if alias_method_chain fails when method is not present, I can do that as well.

**Comment by lzap** [comment]
Right, fixed both.

The tests failed because the node ran out of space. Extended the volume there. We should be fine now.

File: app/lib/puppet_fact_parser_extensions.rb (hunk 1 above)
Comment (stbenjam) [inline app/lib/puppet_fact_parser_extensions.rb:7]: Yea, alias_method_chain does fail silently.  I would think tests should catch it if Foreman plugin "API" changes, since the methods your chaining should get unit tested.  

Discovery only has 2 chained methods, but some plugins have more, so I was thinking this is overkill if you do this every time you modify a Foreman method.

Comment (lzap) [inline app/lib/puppet_fact_parser_extensions.rb:7]: Our tests did not catch it apparently and current tests will unlikely catch this too, because there are no mocha expectations at the moment. We don't need to stub anything for this particular unit test.

If you insist, I can remove this and add two extra tests with an expectation. I just feel this is much more transparent. Someone else can disable or remove the test thinking that it is correct and if that passes review...

Comment (stbenjam) [inline app/lib/puppet_fact_parser_extensions.rb:7]: It's fine as is then

**Comment by stbenjam** [comment]
Looks good to me, still need to test it out

**Comment by stbenjam** [comment]
Merged as 4c018bc136f7db433b3bbcfbdb97cca53bd3e64a, thanks @lzap

**Comment by lzap** [comment]
Many thanks!
